File: packages/typescript/src/features/advanced-capabilities/actions/removeContact.ts
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import {
	findKeywordTermMatch,
	getValidationKeywordTerms,
} from "../../../i18n/validation-keywords.ts";
import { logger } from "../../../logger.ts";
import { removeContactTemplate } from "../../../prompts.ts";
import type { RelationshipsService } from "../../../services/relationships.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { ModelType } from "../../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("REMOVE_CONTACT");
const REMOVE_CONTACT_TERMS = getValidationKeywordTerms(
	"action.removeContact.request",
	{
		includeAllLocales: true,
	},
);

interface RemoveContactXmlResult {
	contactName?: string;
	confirmed?: string;
}

export const removeContactAction: Action = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	examples: (spec.examples ?? []) as ActionExample[][],

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<boolean> => {
		const hasService = !!runtime.getService("relationships");
		const text = message.content.text;
		if (!text) return false;
		const hasIntent = findKeywordTermMatch(text, REMOVE_CONTACT_TERMS);
		return hasService && !!hasIntent;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
	): Promise<ActionResult | undefined> => {
		try {
			const relationshipsService = runtime.getService(
				"relationships",
			) as RelationshipsService;
			if (!relationshipsService) {
				throw new Error("RelationshipsService not available");
			}

			// Build state for prompt composition
			const removeState: State = {
				values: {
					...state?.values,
					message: message.content.text,
					senderName: state?.values?.senderName || "User",
					senderId: message.entityId,
				},
				data: state?.data || {},
				text: state?.text || "",
			};

			const prompt = composePromptFromState({
				state: removeState,
				template: removeContactTemplate,
			});

			const response = await runtime.useModel(ModelType.TEXT_SMALL, {
				prompt,
				stopSequences: [],
			});
			const parsed = parseKeyValueXml<RemoveContactXmlResult>(response);

			if (!parsed?.contactName) {
				logger.warn("[RemoveContact] No contact name provided");
				await callback?.({
					text: "I couldn't determine which contact to remove. Please specify the contact name.",
				});
				return;
			}

			const confirmed = parsed.confirmed?.trim().toLowerCase();
			if (confirmed !== "yes") {
				await callback?.({
					text: `To remove ${parsed.contactName} from your contacts, please confirm by saying "yes, remove ${parsed.contactName}".`,
				});
				return;
			}

			const contacts = await relationshipsService.searchContacts({
				searchTerm: parsed.contactName,
			});

			if (contacts.length === 0) {
				await callback?.({
					text: `I couldn't find a contact named "${parsed.contactName}" in the relationships.`,
				});
				return;
			}

			const contact = contacts[0];

			const removed = await relationshipsService.removeContact(
				contact.entityId,
			);

			if (removed) {
				const responseText = `I've removed ${parsed.contactName} from your contacts.`;
				await callback?.({
					text: responseText,
					actions: ["REMOVE_CONTACT"],
				});

				logger.info(`[RemoveContact] Removed contact ${contact.entityId}`);

				return {
					success: true,
					values: { contactId: contact.entityId },
					data: { success: true },
					text: responseText,
				};
			} else {
				throw new Error("Failed to remove contact");
			}
		} catch (error) {
			logger.error(
				"[RemoveContact] Error:",
				error instanceof Error ? error.message : String(error),
			);
			await callback?.({
				text: "I encountered an error while removing the contact. Please try again.",
				error: error instanceof Error ? error.message : "Unknown error",
			});
		}
	},
};

File: packages/python/elizaos/features/advanced_capabilities/actions/remove_contact.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import REMOVE_CONTACT_TEMPLATE
from elizaos.types import (
    Action,
    ActionExample,
    ActionResult,
    Content,
    ModelType,
)
from elizaos.utils.spec_examples import convert_spec_examples
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )

# Get text content from centralized specs
_spec = require_action_spec("REMOVE_CONTACT")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    return convert_spec_examples(_spec)


@dataclass
class RemoveContactAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        relationships_service = runtime.get_service("relationships")
        return relationships_service is not None

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        from elizaos.features.advanced_capabilities.services.relationships import (
            RelationshipsService,
        )

        relationships_service = runtime.get_service("relationships")
        if not relationships_service or not isinstance(relationships_service, RelationshipsService):
            return ActionResult(
                text="Relationships service not available",
                success=False,
                values={"error": True},
                data={"error": "RelationshipsService not available"},
            )

        state = await runtime.compose_state(message, ["RECENT_MESSAGES", "ENTITIES"])

        prompt = runtime.compose_prompt_from_state(
            state=state,
            template=REMOVE_CONTACT_TEMPLATE,
        )

        response = await runtime.use_model(ModelType.TEXT_SMALL, {"prompt": prompt})
        parsed = parse_key_value_xml(response)

        if not parsed or not parsed.get("contactName"):
            return ActionResult(
                text="Could not determine which contact to remove",
                success=False,
                values={"error": True},
                data={"error": "No contact name provided"},
            )

        contact_name = str(parsed.get("contactName", ""))
        confirmed = str(parsed.get("confirmed", "no")).lower() == "yes"

        if not confirmed:
            response_text = (
                f'To remove {contact_name}, please confirm by saying "yes, remove {contact_name}".'
            )
            if callback:
                await callback(Content(text=response_text, actions=["REMOVE_CONTACT"]))
            return ActionResult(
                text=response_text,
                success=True,
                values={"needsConfirmation": True},
                data={"contactName": contact_name},
            )

        contacts = await relationships_service.search_contacts(search_term=contact_name)

        if not contacts:
            return ActionResult(
                text=f"Could not find a contact named '{contact_name}'",
                success=False,
                values={"error": True},
                data={"error": "Contact not found"},
            )

        contact = contacts[0]
        removed = await relationships_service.remove_contact(contact.entity_id)

        if removed:
            response_text = f"I've removed {contact_name} from your contacts."
            if callback:
                await callback(Content(text=response_text, actions=["REMOVE_CONTACT"]))
            return ActionResult(
                text=response_text,
                success=True,
                values={"contactId": str(contact.entity_id)},
                data={"success": True},
            )
        else:
            return ActionResult(
                text="Failed to remove contact",
                success=False,
                values={"error": True},
                data={"error": "Remove operation failed"},
            )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


remove_contact_action = Action(
    name=RemoveContactAction.name,
    similes=RemoveContactAction().similes,
    description=RemoveContactAction.description,
    validate=RemoveContactAction().validate,
    handler=RemoveContactAction().handler,
    examples=RemoveContactAction().examples,
)
